test/test_framework.py

```diff
@@ -1527,3 +1560,163 @@ def test_named_indicated_all(self, fake_stderr):
+
+
+class DebugHookTests(unittest.TestCase):
+
+    def test_envvar_parsing_missing(self):
+        with patch.dict(os.environ):
+            os.environ.pop('JUJU_DEBUG_AT', None)
+            framework = create_framework(self)
+        self.assertEqual(framework._juju_debug_at, ())
+
+    def test_envvar_parsing_empty(self):
+        with patch.dict(os.environ, {'JUJU_DEBUG_AT': ''}):
+            framework = create_framework(self)
+        self.assertEqual(framework._juju_debug_at, ())
+
+    def test_envvar_parsing_simple(self):
+        with patch.dict(os.environ, {'JUJU_DEBUG_AT': 'hook'}):
+            framework = create_framework(self)
+        self.assertEqual(framework._juju_debug_at, ['hook'])
+
+    def test_envvar_parsing_multiple(self):
+        with patch.dict(os.environ, {'JUJU_DEBUG_AT': 'foo,bar,all'}):
+            framework = create_framework(self)
+        self.assertEqual(framework._juju_debug_at, ['foo', 'bar', 'all'])
+
+    def test_basic_interruption_enabled(self):
+        framework = create_framework(self)
+        framework._juju_debug_at = ['hook']
+
+        publisher = charm.CharmEvents(framework, "1")
+        observer = GenericObserver(framework, "1")
+        framework.observe(publisher.install, observer.callback_method)
+
+        with patch('sys.stderr', new_callable=io.StringIO) as fake_stderr:
+            with patch('pdb.runcall') as mock:
+                publisher.install.emit()
+
+        # Check that the pdb module was used correctly and that the callback method was NOT
+        # called (as we intercepted the normal pdb behaviour! this is to check that the
+        # framework didn't call the callback directly)
+        self.assertEqual(mock.call_count, 1)
+        expected_callback, expected_event = mock.call_args[0]
+        self.assertEqual(expected_callback, observer.callback_method)
+        self.assertIsInstance(expected_event, EventBase)
+        self.assertFalse(observer.called)
+
+        # Verify proper message was given to the user.
+        self.assertEqual(fake_stderr.getvalue(), _BREAKPOINT_WELCOME_MESSAGE)
+
+    def test_actions_are_interrupted(self):
+        test_model = create_model(self)
+        framework = create_framework(self, model=test_model)
+        framework._juju_debug_at = ['hook']
+
+        class CustomEvents(EventSetBase):
+            foobar_action = EventSource(charm.ActionEvent)
+
+        publisher = CustomEvents(framework, "1")
+        observer = GenericObserver(framework, "1")
+        framework.observe(publisher.foobar_action, observer.callback_method)
+        fake_script(self, 'action-get', "echo {}")
+
+        with patch('sys.stderr', new_callable=io.StringIO):
+            with patch('pdb.runcall') as mock:
+                with patch.dict(os.environ, {'JUJU_ACTION_NAME': 'foobar'}):
+                    publisher.foobar_action.emit()
+
+        self.assertEqual(mock.call_count, 1)
+        self.assertFalse(observer.called)
```

Comment:
I think you should use `fake_script_calls` to check that the script was called, no?

Reply:
As we discussed in IRC, that's not really necessary (it's one of the many things that happens before what I really need to test... something I need to fake for the framework to reach the point where happens what is tested, but no more than that).